docs/jep/11859-type-annotations.md

```diff
@@ -0,0 +1,192 @@
+NumPy’s API was developed years before static type checking became part of the Python language, and follows Python’s historic recommendations to use a [duck-typing](https://docs.python.org/3/glossary.html#term-duck-typing)/[EAFP](https://docs.python.org/3/glossary.html#term-eafp) coding style, in which strict type-checking at runtime is discouraged. As a concrete example of this, consider the {func}`numpy.tile` function, which is defined like this:
+
+```python
+def tile(A, reps):
+  try:
+    tup = tuple(reps)
+  except TypeError:
+    tup = (reps,)
+  d = len(tup)
+  ...
+```
+
+Here the *intent* is that `reps` would contain either an `int` or a sequence of `int` values, but the *implementation* allows `tup` to be any iterable.
+When adding annotations to this kind of duck-typed code, we could take one of two routes:
+1. We may choose to annotate the *intent* of the function's API, which here might be something like `reps: Union[int, Sequence[int]]`.
+2. Conversely, we may choose to annotate the *implementation* of the function, which here might look something like `reps: Union[ConvertibleToInt, Iterable[ConvertibleToInt]]` where `ConvertibleToInt` is a special protocol that covers the exact mechanism by which our function converts the inputs to integers (i.e. via `__int__`, via `__index__`, via `__array__`, etc.). Note also here that in a strict sense, `Iterable` is not sufficient here because there are objects in Python that duck-type as iterables but do not satisfy a static type check against `Iterable` (namely, an object that is iterable via `__getitem__` rather than `__iter__`.)
+
+The advantage of #1, annotating intent, is that the annotations are more useful to the user in communicating the API contract; while for the developer the flexibility leaves room for refactoring when necessary. The down-side (particularly for gradually-typed APIs like JAX's) is that it's quite likely that user code exists which runs correctly, but would be flagged as incorrect by a type checker.
+Gradual typing of an existing duck-typed API means that the current annotation is implicitly `Any`, so changing this to a stricter type may present to users as a breaking change.
+
+Broadly speaking, annotating intent better serves Level 1 type checking, while annotating implementation better serves Level 3, while Level 2 is more of a mixed bag (both intent and implementation are important when it comes to annotations in IDEs).
+
+## JAX type annotation roadmap
+
+With this framing (Level 1/2/3) and JAX-specific challenges in mind, we can begin to develop our roadmap for implementing consistent type annotations across the JAX project.
+For JAX type annotation, we will be guided by the following principles:
+
+### Purpose of type annotations
+
+We would like to support full, *Level 1, 2, and 3* type annotation as far as possible. In particular, this means that we should have restrictive type annotations on both inputs and outputs to public API functions.
```

Comment:
> we should have restrictive type annotations on both inputs and outputs

What does "restrictive" mean here? Does it conflict with the section header "inputs should be permissively typed" below?

> to public API functions

I take this to mean that we're not directly concerned with annotating private stuff, for its own sake, as far as this JEP is concerned. Is that right? If so, I like that and encourage it be highlighted more prominently!

Reply:
> If so, I like that and encourage it be highlighted more prominently!

Ideally, with some rationale for why type annotations are less useful for JAX internals :)

Reply:
"restrictive" here means a type that would disallow some inputs. `x: Any` is not restrictive. `x: int` is. When I use "permissive" below, it's referring to the same dimension: we want inputs to be more restrictive than `Any` but still relatively permissive (e.g. `x: Union[int, float, Array]` instead of `x: Any` or `x:Array`) and we want outputs to be as restrictive as possible (e.g. `x: Array` rather than `x: Union[int, float, Array]`)

Reply:
> I take this to mean that we're not directly concerned with annotating private stuff, for its own sake, as far as this JEP is concerned. Is that right? If so, I like that and encourage it be highlighted more prominently!

I think internal annotations are useful as well, and my intent was not to recommend against them in this JEP.